### DISCUSSION TITLE
ERC20 tokens which can be used on zkSycn and enable zkSync bridging

### DISCUSSION THREAD

**Discussion by akyo8** [zkSync-Community-Hub/zksync-developers#39] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

 Hi team
I need to write ERC20 tokens which can be used on zkSycn and enable zkSync bridging. So it would be great if someone give me basic instructions for it.

### What is the specific issue or error you're encountering?

-

### Can you share the error messages or logs you're receiving, if any?

-

### Have you made any recent changes to the contract before encountering this issue?

-

### Are there any external libraries or contracts that your contract interacts with?

-

### Can you provide the relevant portions of your contract code where the issue is occurring?

-

### Have you tried to isolate the problem, and if so, what were the results?

-

### What steps have you already taken to try to resolve the issue?

-

**Comment by uF4No** [comment]
If your token already exists on Ethereum, you can just bridge it with a script that uses the default bridge. You can find [the scripts to deposit and withdraw tokens in this page of the documentation](https://era.zksync.io/docs/reference/concepts/bridging/bridging-asset.html). By using a default bridge, if nobody has sent that token to L2 before, the system will deploy a default ERC20 token contract on L2 behind the scenes, so you don't have to do it manually. Keep in mind that if the original contract on L1 had specific functionalities or methods, those will not be available on L2 if you follow this approach.

On the other hand, you can create your own bridge and deploy the token contract in L2. You can [use L1-L2 messaging for that](https://era.zksync.io/docs/reference/concepts/bridging/l1-l2-interop.html).

**Comment by akyo8** [comment]
Please feel free to check if this approach right.
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import "./interfaces/IERC20.sol";
import "./interfaces/IZkSync.sol";

contract ZkSyncERC20 is IERC20 {
    string public name;
    string public symbol;
    uint8 public decimals;
    uint256 public totalSupply;

    mapping(address => uint256) private balances;
    mapping(address => mapping(address => uint256)) private allowances;

    address private zkSyncAddress;

    constructor(
        string memory _name,
        string memory _symbol,
        uint8 _decimals,
        uint256 _totalSupply,
        address _zkSyncAddress
    ) {
        name = _name;
        symbol = _symbol;
        decimals = _decimals;
        totalSupply = _totalSupply;
        zkSyncAddress = _zkSyncAddress;

        balances[msg.sender] = _totalSupply;
        emit Transfer(address(0), msg.sender, _totalSupply);
    }

    function balanceOf(address account) external view override returns (uint256) {
        return balances[account];
    }

    function transfer(address recipient, uint256 amount) external override returns (bool) {
        require(amount <= balances[msg.sender], "Insufficient balance");

        balances[msg.sender] -= amount;
        balances[recipient] += amount;
        emit Transfer(msg.sender, recipient, amount);

        // If recipient is zkSync address, trigger zkSync transfer
        if (recipient == zkSyncAddress) {
            IZkSync(zkSyncAddress).zkSyncTransfer(address(this), msg.sender, amount);
        }

        return true;
    }

    function transferFrom(address sender, address recipient, uint256 amount) external override returns (bool) {
        require(amount <= balances[sender], "Insufficient balance");
        require(amount <= allowances[sender][msg.sender], "Insufficient allowance");

        balances[sender] -= amount;
        balances[recipient] += amount;
        allowances[sender][msg.sender] -= amount;
        emit Transfer(sender, recipient, amount);

        // If recipient is zkSync address, trigger zkSync transfer
        if (recipient == zkSyncAddress) {
            IZkSync(zkSyncAddress).zkSyncTransfer(address(this), sender, amount);
        }

        return true;
    }

    function approve(address spender, uint256 amount) external override returns (bool) {
        allowances[msg.sender][spender] = amount;
        emit Approval(msg.sender, spender, amount);
        return true;
    }

    function allowance(address owner, address spender) external view override returns (uint256) {
        return allowances[owner][spender];
    }

    function zkSyncTransferFrom(
        address sender,
        address recipient,
        uint256 amount
    ) external {
        require(msg.sender == zkSyncAddress, "Unauthorized");
        require(amount <= balances[sender], "Insufficient balance");

        balances[sender] -= amount;
        balances[recipient] += amount;
        emit Transfer(sender, recipient, amount);
    }
}

**Reply by uF4No** [reply]
I don't understand what are your trying to do with this contract. 

You don't need to use any specific logic or methods for zkSync. You can even just copy an [ERC20 contract from the Openzeppelin wizzard](https://wizard.openzeppelin.com/#erc20) and deploy it to zkSync Era directly and it will just work.

**Comment by akyo8** [comment]
FYI, token is not exist. I am just trying to deploy new ERC20 token contract which can be used on zkSync and enable zkSync bridging.

**Comment by akyo8** [comment]
Also do I just need to add specific functions or implementations to ERC20 standard if token is available on zkSync?

**Comment by akyo8** [comment]
@uF4No Thanks for your reply. Yeah, my ERC20 token is typical form of ERC20 standard from Openzeppelin. Do you mean, I just need to deploy this token contract to zkSync era without change anything? No need to deploy to L1?

**Reply by uF4No** [reply]
if you just want your token on zkSync, yes you don't need to deploy it on Ethereum first.

**Comment by akyo8** [comment]
@uF4No Hi Antonio, Do you mind to give me some instructions?

**Comment by akyo8** [comment]
@uF4No No no. We are in need of token available on Ethereum. But is should be available on zkSync and enable zkSync bridging. So what should be best approach

**Comment by akyo8** [comment]
Hi @uF4No 
How are you doing.
I am trying to follow this guide to deploy this script to enable bridge.
Here is the screenshot but it i
<img width="1829" alt="Screen Shot 2023-08-03 at 12 03 29 AM" src="https://github.com/zkSync-Community-Hub/zkync-developers/assets/28947282/c61fa967-09e2-419c-befc-e03e04dca232">
s saying `Unrecognized task zksync-deploy` . What could be an issue?[](url)

**Reply by uF4No** [reply]
have you installed `hardhat-zksync-deploy` and imported it in the `hardhat.config.ts` file?

**Reply by akyo8** [reply]
Yes, I did. 
Here is the hardhat config.
import { HardhatUserConfig } from "hardhat/config";

import "@matterlabs/hardhat-zksync-deploy";
import "@matterlabs/hardhat-zksync-solc";

import "@matterlabs/hardhat-zksync-verify";

import dotenv from "dotenv";
dotenv.config();

// dynamically changes endpoints for local tests
const zkSyncTestnet =
  process.env.NODE_ENV == "test"
    ? {
        url: "http://localhost:3050",
        ethNetwork: "http://localhost:8545",
        zksync: true,
      }
    : {
        url: "https://zksync2-testnet.zksync.dev",
        ethNetwork: "goerli",
        zksync: true,
        // contract verification endpoint
        verifyURL:
          "https://zksync2-testnet-explorer.zksync.dev/contract_verification",
      };

const config: HardhatUserConfig = {
  zksolc: {
    version: "latest",
    settings: {},
  },
  defaultNetwork: "zkTestnet",

  networks: {
    goerli: {
      url: "https://goerli.infura.io/v3/056f9ac6b8e2455a9492f68a7053e7b0", // The Ethereum Web3 RPC URL (optional).
    },
    zkTestnet: {
      url: "https://testnet.era.zksync.dev", // The testnet RPC URL of zkSync Era network.
      ethNetwork: "goerli", // The Ethereum Web3 RPC URL, or the identifier of the network (e.g. `mainnet` or `goerli`)
      zksync: true,
    },
  },
  solidity: {
    version: "0.8.17",
  },
};

export default config;

**Reply by akyo8** [reply]
But it is not working @uF4No

**Reply by dutterbutter** [reply, marked as answer]
@akyo8 the command is `yarn hardhat deploy-zksync --script <NAME OF FILE>` not `yarn hardhat zksync-deploy --script <FILE>`. I see the error in the docs and will update.

**Comment by akyo8** [comment]
I am just bridging ERC20 token on Goerli to zkTestnet. Also I want to know what other processes we need to do when we go on mainnnet?

**Comment by akyo8** [comment]
https://goerli.etherscan.io/address/0x4341Ff79E0d988D76BfE83CA081624d77f08764E
this is our token contract and we are going to bridging it on zkSync. 
<img width="725" alt="1" src="https://github.com/zkSync-Community-Hub/zkync-developers/assets/28947282/7e9933ec-d572-4346-a2e7-6ce77ba9f3ac">

fyi, deployment proceed after changing hardhat deploy command but still not proceeding here. @dutterbutter

**Comment by akyo8** [comment]
I think this deposit done. But I am not able to find my token available in L2.
0xf90e870f9b4e6c461fe7b64377366ed264e257e455987ab0207a1411bf31ea5d
![Uploading Screen Shot 2023-08-03 at 12.23.36 PM.png…]()

**Reply by uF4No** [reply]
I just checked your [account on the Era testnet explorer](https://goerli.explorer.zksync.io/address/0x82ba3449F70D0BA6D51dCCAd45d0f1a55B5C587A) and you have 5000NAFL tokens. The token address is [0x8D4E861D8C99E473D58A224CFDa896D1E3faB4f4](https://goerli.explorer.zksync.io/address/0x8D4E861D8C99E473D58A224CFDa896D1E3faB4f4#transactions).

Here is the transaction in which your token contract was deployed in L2: https://goerli.explorer.zksync.io/tx/0xd96ca63c826022d21648f05feffa151065625e87e5c67c1819aaaba7df1e2332

**Reply by akyo8** [reply]
Thank you, it is resolved.
